cns/service/main.go

```diff
@@ -709,3 +665,94 @@
 	logger.Printf("CNS exited")
 	logger.Close()
 }
+
+// initializeCRD state
+func IniitalizeCRDState(httpRestService cns.HTTPService, cnsconfig configuration.CNSConfig, exitChan <-chan struct{}) {
+	var requestController requestcontroller.RequestController
+
+	logger.Printf("[Azure CNS] Starting request controller")
+
+	kubeConfig, err := kubecontroller.GetKubeConfig()
+	if err != nil {
+		logger.Errorf("[Azure CNS] Failed to get kubeconfig for request controller: %v", err)
+		return
+	}
+
+	//convert interface type to implementation type
+	httpRestServiceImplementation, ok := httpRestService.(*restserver.HTTPRestService)
+	if !ok {
+		logger.Errorf("[Azure CNS] Failed to convert interface httpRestService to implementation: %v", httpRestService)
+		return
+	}
+
+	// Set orchestrator type
+	orchestrator := cns.SetOrchestratorTypeRequest{
+		OrchestratorType: cns.KubernetesCRD,
+	}
+	httpRestServiceImplementation.SetNodeOrchestrator(&orchestrator)
+
+	// Get crd implementation of request controller
+	requestController, err = kubecontroller.NewCrdRequestController(httpRestServiceImplementation, kubeConfig)
+	if err != nil {
+		logger.Errorf("[Azure CNS] Failed to make crd request controller :%v", err)
+		return
+	}
+
+    // initialize the ipam pool monitor
+	httpRestServiceImplementation.IPAMPoolMonitor = ipampoolmonitor.NewCNSIPAMPoolMonitor(httpRestServiceImplementation, requestController)
+
+	err = requestController.InitRequestController()
+	if err != nil {
+		logger.Errorf("[Azure CNS] Failed to initialized cns state :%v", err)
+		return
+	}
+
+	//Start the RequestController which starts the reconcile loop
+	go func() {
+		for {
+			if err := requestController.StartRequestController(exitChan); err != nil {
+				logger.Errorf("[Azure CNS] Failed to start request controller: %v", err)
+				// retry to start the request controller
+				// todo: add a CNS metric to count # of failures
+			} else {
+				logger.Printf("[Azure CNS] Exiting RequestController")
+				return
+			}
+
+			// Retry after 1sec
+			time.Sleep(time.Second)
+		}
+	}()
+
+	for {
+		if requestController.IsStarted() {
+			logger.Printf("RequestController is started")
+			break
+		}
+
+		logger.Printf("Waiting for requestController to start...")
+		time.Sleep(time.Millisecond * 500)
+	}
+
+	ctx := context.Background()
```

Comment:
At some point we need to use contexts properly, such as adding a cancel func or timeout. However, this PR doesn't need to address this, since there are `context.Background()`s and `context.TODO()`s all throughout the CNS code, so we probably need one PR just to address context.

Reply:
Agreed

---
In reply to: [585724212](https://github.com/Azure/azure-container-networking/pull/809#discussion_r585724212) [](ancestors = 585724212)